### PR TITLE
fix(v2,sdk.v2): Remove need for base image root access to write artifacts

### PR DESCRIPTION
**Pull request by judahrand**
**Description of your changes:**
The changes in this PR are based on @Bobgy's design [here](https://github.com/kubeflow/pipelines/issues/5673#issuecomment-901045901).

Custom base images may be built specifically to avoid granting root privileges. Without these privileges the local mount points cannot be created if they live at the root of the filesystem. `/gcs` must remain where it is for now as it is currently a contract for KFP v2 python component wrappers. Therefore, this directory is also mounted as an `emptyDir` so that it already exists and is accessible to all users.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/v2/components/types/artifact_types.py [hunk 1]
```diff
@@ -19,9 +19,9 @@
 import os
 from typing import Dict, Generic, List, Optional, Type, TypeVar, Union
 
-_GCS_LOCAL_MOUNT_PREFIX = '/gcs/'
-_MINIO_LOCAL_MOUNT_PREFIX = '/minio/'
-_S3_LOCAL_MOUNT_PREFIX = '/s3/'
+_GCS_LOCAL_MOUNT_PREFIX = '/tmp/gcs/'
```
Comment (chensun) [inline sdk/python/kfp/v2/components/types/artifact_types.py:22]: This needs to be `/gcs/` for [GCS Fuse](https://cloud.google.com/storage/docs/gcs-fuse) support on Vertex Pipelines.
The content you write to the local path `/gcs/my_project/bucket/blob` would be automatically synced to `gs://my_project/bucket/blob`.

Comment (judahrand) [reply sdk/python/kfp/v2/components/types/artifact_types.py:22]: Do you have an alternate solution for being able to execute in non-root images?

**Comment by judahrand** [comment]
I've opened a similar issue with Vertex Pipelines:
https://issuetracker.google.com/issues/199297769

File: sdk/python/kfp/v2/components/types/artifact_types.py (hunk 1 above)
Comment (chensun) [reply sdk/python/kfp/v2/components/types/artifact_types.py:22]: I think it is an implementation decision by Vertex Pipelines backend. Following that, we probably have multiple places in code where we rely on this conversation between local path and gs uri. 
Another example:
https://github.com/kubeflow/pipelines/blob/3a621fb577756171c726a423ce68a513b9053b8f/v2/component/launcher.go#L655

**Comment by chensun** [comment]
> I've opened a similar issue with Vertex Pipelines:
> https://issuetracker.google.com/issues/199297769

Thank you for opening the issue. I'll pass it along to the Vertex Pipelines team.

Meanwhile, I am curious is there any specific reason you want to avoid running container as root user? Or it is an image out of your control?

File: sdk/python/kfp/v2/components/types/artifact_types.py (hunk 1 above)
Comment (judahrand) [reply sdk/python/kfp/v2/components/types/artifact_types.py:22]: This PR does also change those.

But you're probably right that it's actually probably a problem closer to how Vertex use `gcsfuse` and the file permissions which are set there in my case.

However, do you not think that trying to create these directories on at the root of the filesystem is an unnecessary hurdle in the way of non-root containers?

**Comment by judahrand** [comment]
> > I've opened a similar issue with Vertex Pipelines:
> > https://issuetracker.google.com/issues/199297769
> 
> Thank you for opening the issue. I'll pass it along to the Vertex Pipelines team.
> 
> Meanwhile, I am curious is there any specific reason you want to avoid running container as root user? Or it is an image out of your control?

It is an image which we also use in our GKE cluster and our policy there is to use non root containers wherever possible. It would be possible to create a different image for Vertex... But that seems unnecessary.

**Comment by chensun** [comment]
> > I've opened a similar issue with Vertex Pipelines:
> > https://issuetracker.google.com/issues/199297769
> 
> Thank you for opening the issue. I'll pass it along to the Vertex Pipelines team.
> 
> Meanwhile, I am curious is there any specific reason you want to avoid running container as root user? Or it is an image out of your control?



> > > I've opened a similar issue with Vertex Pipelines:
> > > https://issuetracker.google.com/issues/199297769
> > 
> > 
> > Thank you for opening the issue. I'll pass it along to the Vertex Pipelines team.
> > Meanwhile, I am curious is there any specific reason you want to avoid running container as root user? Or it is an image out of your control?
> 
> It is an image which we also use in our GKE cluster and our policy there is to use non root containers wherever possible. It would be possible to create a different image for Vertex... But that seems unnecessary.

Got it. 
Maybe something related
- Kubernetes is deprecating Docker as a container runtime after v1.20. Part of the reason is for security: https://www.kubeflow.org/docs/components/pipelines/installation/choose-executor/#docker-executor
- And KFP is offering `emissary` executor, which will become the default in the future: https://github.com/kubeflow/pipelines/issues/5718

I wonder if these changes may address your concerns behind the policy of avoiding root containers. 

That said, I'll still pass your concern to the team, and see if we could make a change.

**Comment by judahrand** [comment]
Unfortunately, I don't set the rules 😉

I also suspect that we'll continue to see many K8s installations using the Docker shim for quite a while to come. 

I appreciate that there is progress being made on the security side of containerization which may make non-root less important but it still just strikes me as an odd failure mode (especially as it isn't documented). 

Thank you very much for passing the issue along - I can't imagine a small file permission tweak is that big a change to enable non-root 😛

**Comment by Bobgy** [comment]
FYI, there were a similar request raised for v2 compatible mode: https://github.com/kubeflow/pipelines/issues/5673

**Comment by judahrand** [comment]
> FYI, there were a similar request raised for v2 compatible mode: #5673

Ah! I missed this - might as well close this in that case. I'll mention it in that issue too so as not to lose the reference to the Vertex issue with Google.

**Comment by judahrand** [comment]
Here is where some some these changes are needed I think. https://github.com/kubeflow/pipelines/blob/a0b18eb9e8da25c7b81b0b75cf0f39297c809d20/v2/compiler/container.go#L134

**Comment by judahrand** [comment]
@Bobgy I've altered this PR to try to tackle #5673. I'd appreciate any feedback - I may well be going in very much the wrong direction?

**Comment by Bobgy** [comment]
@judahrand thank you for this PR that fixes non root container for v2 compatible mode!

Note, because compiled v2 python component yamls need to be recompiled to get the updated path. We need to be careful how to roll this out.

@chensun do you have any suggestions?

My impression is that compiling v2 python components to component yaml is a very corner case usage. We might be able to just roll this out without any backwards compatible cautions.

**Comment by judahrand** [comment]
@Bobgy The cause of the `kubeflow-pipelines-samples-v2` failures may well be very obvious but I'm not able to debug them as I don't have access to view the dashboard. Not sure if you're willing to (temporarily?) give me access (https://github.com/kubeflow/pipelines/blob/master/v2/test/README.md#how-to-access-the-kfp-ui-running-these-tests)?

**Comment by chensun** [comment]
> @chensun do you have any suggestions?
> 
> My impression is that compiling v2 python components to component yaml is a very corner case usage. We might be able to just roll this out without any backwards compatible cautions.

v2 Python components is relatively new, so it's probably rare for users to export them into component yaml at this moment, but I don't think it would be a corner case going forward.

**Comment by chensun** [comment]
The PR initially was trying to change the artifacts mounting point `/gcs/...`. Isn't that still needed until you can use non-root containers?
BTW, the title and description of this PR need to be updated.

**Comment by judahrand** [comment]
> The PR initially was trying to change the artifacts mounting point `/gcs/...`. Isn't that still needed until you can use non-root containers?
> BTW, the title and description of this PR need to be updated.

The changes in this PR are based on @Bobgy's design (here)[https://github.com/kubeflow/pipelines/issues/5673#issuecomment-901045901]. The `/gcs/` directory is mounted as an `emptyDir` volume as so should be accessible to non-root users.

The fact that `gcsfuse` seems to mess with the file permissions may still need to be addressed in Vertex Pipelines but this should solve the issue for non-Vertex use cases.

**Comment by judahrand** [comment]
> `go: github.com/argoproj/argo-workflows/v3@v3.1.1: Get "https://proxy.golang.org/github.com/argoproj/argo-workflows/v3/@v/v3.1.1.mod": net/http: TLS handshake timeout`

Looks like the `kubeflow-pipelines-v2-go-test` failure isn't me?

/retest

**Comment by Bobgy** [comment]
/assign @capri-xiyue 
Can you help with review? This is a frequently requested feature.

**Comment by capri-xiyue** [comment]
I'm wondering shall we wait for the fix of this until we release KFP V2 and get more feedback from users?

**Comment by judahrand** [comment]
> I'm wondering shall we wait for the fix of this until we release KFP V2 and get more feedback from users?

Is there a specific reason for this? It seems like there's not too much downside to this and implementing it sooner rather than later will mean fewer users need to recompile their `yaml`?

**Comment by capri-xiyue** [comment]
> need to recompile their `yaml`

Why users need to recompile their yaml if we fix this after we release v2?

**Comment by judahrand** [comment]
> @judahrand thank you for this PR that fixes non root container for v2 compatible mode!
> 
> Note, because compiled v2 python component yamls need to be recompiled to get the updated path. We need to be careful how to roll this out.
> 
> @chensun do you have any suggestions?
> 
> My impression is that compiling v2 python components to component yaml is a very corner case usage. We might be able to just roll this out without any backwards compatible cautions.

@capri-xiyue 
To get the updated paths.

**Comment by capri-xiyue** [comment]
> ed8f635#diff-dd6640ee36f4d49a93ee28d4ef574e3e79f28a4561691e3e716700bb9cf2c125

I think the sdk you changed only affects the v2 compatible mode. It does not affect v2 mode.
This pr need to change the v2 golang code, but I think they are implementation details and they don't affect users.

**Comment by capri-xiyue** [comment]
> > I'm wondering shall we wait for the fix of this until we release KFP V2 and get more feedback from users?
> 
> Is there a specific reason for this? It seems like there's not too much downside to this and implementing it sooner rather than later will mean fewer users need to recompile their `yaml`?



I'm wondering maybe we should allow user to customize access? In addition, is it possible that user will need root access in 

> > I'm wondering shall we wait for the fix of this until we release KFP V2 and get more feedback from users?
> 
> Is there a specific reason for this? It seems like there's not too much downside to this and implementing it sooner rather than later will mean fewer users need to recompile their `yaml`?

Now we use a workaround for gcs path. I'm wondering shall we wait for the v2 release to gather more user feedback and then Vertex and KFP can discuss a solution that works for both sides.

**Comment by judahrand** [comment]
> Now we use a workaround for gcs path. I'm wondering shall we wait for the v2 release to gather more user feedback and then Vertex and KFP can discuss a solution that works for both sides.

As far as I can tell from the outside and fiddling around in Vertex they're just using `gcsfuse` to mount the `/gcs` directory. According to (this)[https://github.com/GoogleCloudPlatform/gcsfuse#invoking-gcsfuse] they:
1. can mount that directory anywhere they like
2. should not be as the user which will be using the directory `gcsfuse` (not `root` which is what it seems they are doing now). 

From an outside perspective this just seems like a bug/error on their side. But I can't get in reliable contact with anyone on the inside so hard to tell for certain...

**Comment by judahrand** [comment]
> I'm wondering maybe we should allow user to customize access? In addition, is it possible that user will need root access in

Effectively, this would allow the user to decide whether to run as `root` or not by using a Dockerfile which does or does not declare a `USER`? KFP could still deploy as `root` for now and the user can handle what `user` runs in the image?

**Comment by capri-xiyue** [comment]
@judahrand  The community discussed with Vertex side. The community decided to remove the need for base image root access to write artifacts for KFP now. It looks like the sample-v2-test is failing. Can you help fix it? In addition, I'm wondering whether you can see the error logs of the sample-v2-test. Can you help share the screenshot of the error logs you saw? I think you need to be in https://github.com/kubeflow/internal-acls/blob/master/google_groups/groups/ci-team.yaml to see the error logs in test infra.

**Comment by judahrand** [comment]
> @judahrand The community discussed with Vertex side. The community decided to remove the need for base image root access to write artifacts for KFP now. It looks like the sample-v2-test is failing. Can you help fix it? In addition, I'm wondering whether you can see the error logs of the sample-v2-test. Can you help share the screenshot of the error logs you saw? I think you need to be in https://github.com/kubeflow/internal-acls/blob/master/google_groups/groups/ci-team.yaml to see the error logs in test infra.

@capri-xiyue I'm keen to help fix it but I'm not able to see the logs of that test unfortunately. I opened a PR in `internal-acls` that should give me the access. But I'm blocked on this until I have that access.

**Comment by capri-xiyue** [comment]
> > @judahrand The community discussed with Vertex side. The community decided to remove the need for base image root access to write artifacts for KFP now. It looks like the sample-v2-test is failing. Can you help fix it? In addition, I'm wondering whether you can see the error logs of the sample-v2-test. Can you help share the screenshot of the error logs you saw? I think you need to be in https://github.com/kubeflow/internal-acls/blob/master/google_groups/groups/ci-team.yaml to see the error logs in test infra.
> 
> @capri-xiyue I'm keen to help fix it but I'm not able to see the logs of that test unfortunately. I opened a PR in `internal-acls` that should give me the access. But I'm blocked on this until I have that access.

I lgtm for the pr yoo opened in  `internal-acls`

**Comment by Bobgy** [comment]
Sorry for the confusion, the test was indeed flaky recently and I think I just resolved it.

**Comment by judahrand** [comment]
Is there anything else here that needs to be done to make this mergeable?

**Comment by Bobgy** [comment]
Thank you @judahrand!
/lgtm

**Comment by chensun** [comment]
/approve

Thanks!

**Comment by dandawg** [comment]
This looks to be close to the finish line. What work is still needed to get it across, and do we have a timeline?

**Comment by zijianjoy** [comment]
Hello @judahrand , would you like to resolve the merge conflict and run test again? This PR has been approved so we don't expect much changes required beyond that.
